### ISSUE TITLE
cannot load or save in linux build

### ISSUE THREAD

**Issue by stephematician** [mefistotelis/swars#43]
Save and load games aren't functioning on linux - i suspect there's some issue with directories not being created or the paths not being correctly inferred.

Here's the error.log (release build)...

```
bflibrary ver 0.1.2.0 (standard release)
LOG CREATED  @ 00:12:42  25-11-2022

Error: Bflib: LbFileOpen: file does not exist: "data/tex06.dat"
Error: Bflib: LbMemoryAllocLow: failed memory allocation of 256 bytes
Error: Bflib: LbDataLoadAll: *VESA: cannot allocate memory
Warning: Bflib: LbScreenUpdateIcon: cannot set icon: image load failed: Couldn't open swars_icon.png
Warning: Bflib: LbScreenUpdateIcon: cannot set icon: image load failed: Couldn't open swars_icon.png
Warning: Bflib: LbScreenUpdateIcon: cannot set icon: image load failed: Couldn't open swars_icon.png
Warning: Bflib: LbScreenUpdateIcon: cannot set icon: image load failed: Couldn't open swars_icon.png
Error: Bflib: LbFileOpen: file does not exist: "data\peepdefs.dat"
Error: Bflib: LbFileOpen: file does not exist: "qdata/savegame/synwars0.sav"
Error: Bflib: LbFileOpen: file does not exist: "qdata/savegame/synwars1.sav"
Error: Bflib: LbFileOpen: file does not exist: "qdata/savegame/synwars2.sav"
Error: Bflib: LbFileOpen: file does not exist: "qdata/savegame/synwars3.sav"
Error: Bflib: LbFileOpen: file does not exist: "qdata/savegame/synwars4.sav"
Error: Bflib: LbFileOpen: file does not exist: "qdata/savegame/synwars5.sav"
Error: Bflib: LbFileOpen: file does not exist: "qdata/savegame/synwars6.sav"
Error: Bflib: LbFileOpen: file does not exist: "qdata/savegame/synwars7.sav"
```

**Comment by mefistotelis** [comment]
Yeah, user data paths are not well unified.

There's the old `game_transform_path`, and the new `SyndFileNameTransform`. The old one is probably still used for something.
There's the old define `SAVEGAME_PATH` but there is also a global var `game_dir_savegame`.

It is possible that there is a disagreement between the calls on where the saved games should be - then the `LbDirectoryMake(game_dir_savegame, true);` might be creating a path which isn't always used.

Or maybe you just didn't gave proper permissions to the folder.

**Comment by stephematician** [comment]
Permission are ok. It looks like something to do with file name conventions.

When I open the load screen it looks like the game searches for `qdata/savegame/synwars%d.sav` to fill out the save/load slots; but the files in the directory take their name from whatever was entered in the menu when it was saved, e.g. `foo.sav` if I entered `foo`.

I'll keep debugging, but if you see this and have ideas in the meantime, that's good, too.

**Comment by mefistotelis** [comment]
You might be misinterpreting something. Saves are either `synwarsN.sav` or `swarsNNN.sav` (when above 8). Only INI has a user-provided name. Just verified this on my win32 build.

For the `game_transform_path` - that's not an issue as well, as from what I see this old method is only used when opening SMK movies.

You should try printing file names after they go through `SyndFileNameTransform` (which the `lbFileNameTransform` points to).

**Comment by stephematician** [comment]
Thanks for the info; you were right, I was very tired when I was looking at the file names.

First thing I figured out was that it always writes to `swars219.sav` - regardless of the slot I select. This file won't appear on the list because the menu (in game) only goes up to the 99th element, `swars098.sav`).

I looked at the call stack from lbFileOpen ... and sighed; it is called from line 42303 in swars.sx. Working backwards I found that at line 42282 of `swars.sx` the value 220 is moved from `($esp + 0x2c)`

```asm
	jump_349e0:
		xor    %edx,%edx
		mov    0x2c(%esp),%dl
; (gdb) print $esp
; = 0xffffca84
; (gdb) print *(int *)(($esp) + 0x2c)
; = 220
```

And then shortly after the prefix for the filename is constructed via sprintf using this value (minus 1)

```asm
	jump_349ff:
		lea    0x8(%esp),%eax
		push   %eax
		call   ac_sprintf
; (gdb) print (char *)($esp + 0xc)
; = 0xffffca84 "qdata/savegame/swars219.sav"
```

Working backwards I found that the 220 value was stored at the address (which was eventually used to make the filename) at the start of `ASM_save_game_write` ...

```asm
GLOBAL_FUNC(ASM_save_game_write)	;/* 0x3422C */
		push   %ebx
		push   %ecx
		push   %esi
		push   %edi
		push   %ebp
		sub    $0x30,%esp
		mov    %al,0x2c(%esp)
; (gdb) print $esp
; = 0xffffca84
; (gdb) print (unsigned char)$al
; = 220 '\334'
```

The call stack wasn't helpful, but I found that `ASM_save_game_write` was being called from`save_game_slot`, specifically line 233974 of `swars.sx`:

```asm
		mov    data_1551d8,%al
		call   ac_save_game_write
; (gdb) info registers 
; eax            0x1                 1
; ecx            0x0                 0
; edx            0x56762748          1450583880
; ebx            0x0                 0
; esp            0xffffcaf0          0xffffcaf0
```

which goes to `save_game_write` in `game.c` first before being passed on to `ASM_save_game_write`. So the `al` register looks fine here (has the value 1, as expected), but something happens in `game.c`.  Stepping through from the call from `save_game_slot` ... it seems as though the eax register is immediately altered when entering the `save_game_write` function scope:

```sh
(gdb) s
# save_game_write () at ../../src/game.c:2616
# 2616	{
(gdb) info registers
# eax            0x1                 1
(gdb) n
eax            0x56701cdc          1450187996
```

I'm not clever enough to know why the register value changes when entering the scope of `save_game_write` from `game.c` - any ideas?

**Comment by mefistotelis** [comment]
Good debug!
`ASM_save_game_write` clearly requires save game slot in `al` (1st parameter is in eax in the Watcom calling convention).
I made a mistake during rewrite. And on win32 I was just lucky, as eax contained what it should.

You need to modify wrapper entry so that the C function gets int parameter from asm.
And modify C declaration of the `save_game_write` so that it accepts one parameter and places it in eax (mnemonic for eax in the extended asm syntax is "a").

**Comment by stephematician** [comment]
RE: second argument, yep, good catch! I noticed that `edx` was set, too, and it seems you are correct re: what it addresses. Save game issue seems fixed now with 68e27012cf710ec4bc8f793796dd4613321867d5 - thanks!
